### PR TITLE
Introduce a freshness indicator for the goal screen

### LINKED ISSUE #246: Loading indicator for goals
```[tasklist]
### Desiderata
- [ ] Does this replicate in the current version of the app?
- [ ] TBD
```

Request [from the forum](https://forum.beeminder.com/t/ios-app-loading-indicator-for-goals/5595):

> When I launch the iOS app, it often gives me a scare. If I’ve entered data for one of my goals outside the app, and that goal would have otherwise derailed, then it shows a big red LOST when the app first opens. It sends a stab of fear through me, until the app updates its data and refreshes to show no, my goal is still on track. Phew.
>
> This happens almost every time I open the app. Can you add a neutral “loading” indicator when it launches, so it doesn’t just assume I’ve derailed? Or maybe just for the goals that would have derailed, wait until you’ve actually refreshed the data before showing that scary red indicator.

The app (Version 5.8) does display a textual representation of the time the list of goals was last updated.

## Cognata

* https://github.com/beeminder/BeeSwift/blob/120beaaf96c06082280141f9acac09775a4d2091/BeeSwift/GalleryViewController.swift#L345

_**Verbata**: microcopy, status indicators, freshness / staleness of graphs, scaring poor users with spurious derailment indicators,_

### PR DESCRIPTION
**Pull request by krugerk**
## Summary
In line with #246, this provides an indicator of freshness on the goal screen. It reuses the one from the gallery screen.

## Media

https://github.com/user-attachments/assets/b2549822-e1b0-42de-9dea-f451bee4fad4

## Validation
Ran app in simulator
Viewed both gallery and goal, navigating back and forth.
Sent the app to background, back into foreground, varying sometimes with the gallery open, sometimes with a goal open.
Also watched the last updated since label be updated while just leaving the app open with the goal view showing

Fixes #246

### CODE REVIEW

File: BeeKit/Managers/GoalManager.swift [hunk 1]
```diff
@@ -25,6 +25,9 @@ public actor GoalManager {
     private nonisolated let currentUserManager: CurrentUserManager
 
     public var goalsFetchedAt : Date? = nil
+    
+    // id, not goalname (slug)
+    private var lastFetchedByGoalId: [String: Date?] = [:]
```
Comment (theospears) [inline BeeKit/Managers/GoalManager.swift:30]: I think `Goal.lastModifiedLocal` may already be tracking the value you want?

Comment (krugerk) [reply BeeKit/Managers/GoalManager.swift:30]: I was looking for when the app last fetched the data from the server. With Core Data, theoretically we could be editing objects locally (offline mode?) and at some point syncing them with the server.

Given _local_ in the name and the description of the attribute,
https://github.com/beeminder/BeeSwift/blob/745cf9875ed6b0303c46d5d8e4c0a2d44796040d/BeeKit/Model/Goal.swift#L66-L67
I had the impression this is meant to reflect when the goal was last touched locally and does not necessarily reflect also when the goal was last fetched from the server. I see this date is being set whenever updateToMatch is being called which is being called after a fetch of user/goal from the backend.

Comment (krugerk) [reply BeeKit/Managers/GoalManager.swift:30]: I see it, `goal.lastModifiedLocal` is also set with Date.now when the object is created (pulled from the db). In this case the lastModifiedLocal does not reflect how fresh the data is (when last pulled from server).

Comment (theospears) [reply BeeKit/Managers/GoalManager.swift:30]: > goal.lastModifiedLocal is also set with Date.now when the object is created (pulled from the db). 

I don't believe this is true, at the very least it's not the intended behavior. While there is a constructor which sets the value, I don't believe this is the constructor which is invoked when materializing from the database.

Comment (krugerk) [reply BeeKit/Managers/GoalManager.swift]: Alright, you convinced me!

Comment (theospears) [reply BeeKit/Managers/GoalManager.swift]: Yeah, I thought the explicit list all attributes constructors might be useful for tests. If we had more tests.

File: BeeKit/Managers/GoalManager.swift [hunk 2]
```diff
@@ -72,19 +72,20 @@ public actor GoalManager {
     public func refreshGoal(_ goalID: NSManagedObjectID) async throws {
         let goal = try modelContext.existingObject(with: goalID) as! Goal
 
-        let responseObject = try await requestManager.get(url: "/api/v1/users/\(currentUserManager.username!)/goals/\(goal.slug)?datapoints_count=5", parameters: nil)
+        let responseObject = try await requestManager.get(url: "/api/v1/users/\(goal.owner.username)/goals/\(goal.slug)?datapoints_count=5", parameters: nil)
```
Comment (theospears) [inline BeeKit/Managers/GoalManager.swift:75]: This change seems fine and we can merge with it in, though I'll once again request to minimize unrelated cleanup (there are no feature related changes to this file) in feature PRs.

File: BeeKit/Managers/RequestManager.swift [hunk 3]
```diff
@@ -39,6 +39,8 @@ public class RequestManager {
             urlWithSubstitutions = urlWithSubstitutions.replacingOccurrences(of: "{username}", with: username)
         }
 
+        logger.debug("\(#function) | \(method.rawValue) | \(urlWithSubstitutions)")
```
Comment (theospears) [inline BeeKit/Managers/RequestManager.swift:42]: Again, unrelated change.

As a reviewer it can be hard to know if changes like this are deliberate, or were intended as temporary debugging measures, and weren't cleaned up before opening the PR.

File: BeeSwift/GoalViewController.swift [hunk 4]
```diff
@@ -24,6 +24,14 @@ class GoalViewController: UIViewController,  UIScrollViewDelegate, DatapointTabl
 
     let goal: Goal
 
+
+    private var lastFetched: Date? {
```
Comment (theospears) [inline BeeSwift/GoalViewController.swift:28]: This doesn't seem to be a real variable. By this I mean we don't really keep track of a value in it - the only reads are within `updateLastUpdatedLabel`, which could just as happily be reading directly from `goal.lastModifiedLocal`

[Do] Let's remove it, and just call `updateLastUpdatedLabel()` in the places we were writing to it.

File: BeeSwift/GoalViewController.swift [hunk 5]
```diff
@@ -63,9 +72,20 @@ class GoalViewController: UIViewController,  UIScrollViewDelegate, DatapointTabl
...
 
+        self.view.addSubview(self.timeElapsedView)
+        self.timeElapsedView.snp.makeConstraints { (make) -> Void in
+            make.top.equalTo(self.view.safeAreaLayoutGuide.snp.topMargin)
+            make.left.equalTo(0)
+            make.right.equalTo(0)
+        }
+        
+        self.updateLastUpdatedLabel()
+        Timer.scheduledTimer(timeInterval: 60, target: self, selector: #selector(GoalViewController.updateLastUpdatedLabel), userInfo: nil, repeats: true)
```
Comment (theospears) [inline BeeSwift/GoalViewController.swift:83]: [Consider] Wouldn't this timer make more sense being part of the `FreshnessIndicatorView` ?

**Review by theospears — CHANGES_REQUESTED** [review]
Overall approach looks good, and I like the idea of extracting `FreshnessIndicatorView` as its own component. I added some requests for local changes.

When testing, could you also verify the time since update changes when you just leave the app open with the goal view showing?

File: BeeKit/Managers/RequestManager.swift (hunk 3 above)
Comment (krugerk) [reply BeeKit/Managers/RequestManager.swift:42]: It is not necessary for the MR itself though it was deliberate to facilitate development.

I am of the opinion that one should be able to follow log statements back to figure out what went wrong. The logging in the app is not yet to that point.

**Review by theospears — CHANGES_REQUESTED** [review]
Still waiting on changes to GoalViewController

File: BeeSwift/GoalViewController.swift (hunk 5 above)
Comment (krugerk) [reply BeeSwift/GoalViewController.swift:83]: The reason I placed it here is because this is also a Controller whereas the Freshness one is just a view.

File: BeeKit/Managers/GoalManager.swift (hunk 2 above)
Comment (krugerk) [reply BeeKit/Managers/GoalManager.swift]: removed from this PR

File: BeeSwift/GoalViewController.swift [hunk 6]
```diff
@@ -24,6 +24,14 @@ class GoalViewController: UIViewController,  UIScrollViewDelegate, DatapointTabl
...
+    private var lastFetched: Date? {
+        didSet {
+            updateLastUpdatedLabel()
+        }
+    }
+    
+    private let timeElapsedView = FreshnessIndicatorView()
     fileprivate var goalImageView = GoalImageView(isThumbnail: false)
     fileprivate var datapointTableController = DatapointTableViewController()
     fileprivate var dateTextField = UITextField()
```
Comment (krugerk) [inline BeeSwift/GoalViewController.swift:30]:
```suggestion
    fileprivate let goalImageView = GoalImageView(isThumbnail: false)
    fileprivate let datapointTableController = DatapointTableViewController()
    fileprivate let dateTextField = UITextField()
```

File: BeeSwift/Gallery/GalleryViewController.swift [hunk 7]
```diff
@@ -22,8 +22,7 @@ class GalleryViewController: UIViewController, UICollectionViewDelegateFlowLayou
 
     var collectionView :UICollectionView?
     var collectionViewLayout :UICollectionViewFlowLayout?
-    let lastUpdatedView = UIView()
-    let lastUpdatedLabel = BSLabel()
+    private let freshnessIndicator = FreshnessIndicatorView()
     let cellReuseIdentifier = "Cell"
     var deadbeatView = UIView()
     var outofdateView = UIView()
```
Comment (krugerk) [inline BeeSwift/Gallery/GalleryViewController.swift:28]:
```suggestion
    private let deadbeatView = UIView()
    private let outofdateView = UIView()
```

File: BeeKit/Managers/RequestManager.swift (hunk 3 above)
Comment (krugerk) [reply BeeKit/Managers/RequestManager.swift]: removed from this PR

File: BeeSwift/GoalViewController.swift (hunk 4 above)
Comment (krugerk) [reply BeeSwift/GoalViewController.swift]: done

**Review by theospears — APPROVED** [review]

**Comment by theospears** [comment]
LGTM, merged!

File: BeeSwift/GoalViewController.swift (hunk 4 above)
Comment (krugerk) [reply BeeSwift/GoalViewController.swift]: unfortunately that has the side effect of the app crashing when accessing goal.lastModifiedLocal at GoalViewController.updateLastUpdatedLabel upon signing into the app.

1. GoalManager.performPostGoalUpdateBookkeeping posts notification goalsUpdated
2. GoalViewController.onGoalsUpdatedNotification executes
3. calling its updateInterfaceToMatchGoal
4. GoalViewController.updateInterfaceToMatchGoal calls, among others, updateLastUpdatedLabel
5. GoalViewController.updateLastUpdateLabel now accesses self.goal.lastModifiedLocal , which at the time, is still part of an empty NSManagedObject Goal:

```swift
<Goal: 0x600002148aa0> (entity: Goal; id: 0xblahblahblah9494 <x-coredata://blahblah-F9F9-4FB3-BD1B-blahblah/Goal/p385>; data: {
    alertStart = nil;
    autodata = nil;
    data = "<relationship fault: 0x600000383740 'data'>";
    deadline = nil;
    graphUrl = nil;
    healthKitMetric = nil;
    hhmmFormat = nil;
    id = nil;
    initDay = nil;
    lastModifiedLocal = nil;
    lastTouch = nil;
    leadTime = nil;
    limSum = nil;
    owner = nil;
    pledge = nil;
    queued = nil;
    recentData = "<relationship fault: 0x6000003829a0 'recentData'>";
    safeBuf = nil;
    safeSum = nil;
    slug = nil;
    thumbUrl = nil;
    title = nil;
    todayta = nil;
    urgencyKey = nil;
    useDefaults = nil;
    won = nil;
    yAxis = nil;
})
```

Comment (theospears) [reply BeeSwift/GoalViewController.swift]: It seems likely that versions of this bug will end up existing, or be easy to introduce, all across GoalView. Do you have ideas for why the GoalViewController even exists while in this state? We're certainly not showing it.

Comment (theospears) [reply BeeSwift/GoalViewController.swift]: Tracking this as #590
